vtrace/util.py

```diff
@@ -76,3 +78,220 @@ def fireLocalNotifiers(self, event, trace):
         the traces. (used to locally bump notifiers)
         """
         self.dnotif.notify(event, trace)
+
+
+def emuFromTrace(trace):
+    '''
+    Produce an envi emulator for this tracer object.
+    '''
+    import envi
+    import envi.memory as e_memory
+    arch = trace.getMeta('Architecture')
+    plat = trace.getMeta('Platform')
+    amod = envi.getArchModule(arch)
+    emu = amod.getEmulator()
+    [emu.setMeta(key, val) for key, val in trace.metadata.items()]
+
+    # could use {get,set}MemorySnap if trace inherited from MemoryObject
+    for va, size, perms, fname in trace.getMemoryMaps():
+        try:
+            # So linux maps in a PROT_NONE page for efficient library sharing, so we have to take that into account
+            if (not perms & e_memory.MM_READ):
+                continue
+            if plat == 'linux' and fname in ['[vvar]']:
+                continue
+            bytez = trace.readMemory(va, size)
+            emu.addMemoryMap(va, perms, fname, bytez)
+        except vtrace.PlatformException:
+            print('failed to map: 0x{:x} into emu'.format(va, size))
+            continue
+
+    rsnap = trace.getRegisterContext().getRegisterSnap()
+    emu.setRegisterSnap(rsnap)
+
+    if plat == 'windows':
+        psize = trace.getPointerSize()
+        # capture PEB and TIB
+        peb = trace.getMeta('PEB')
+        tebs = dict(trace.win32threads)
+        # test and make sure:
+        for threadid, teb in tebs.items():
+            assert(trace.readMemoryPtr(teb + (psize * 12)) == peb)
+            assert(trace.readMemoryPtr(teb + (psize * 9)) == threadid)
+
+        emu.setMeta('PEB', peb)
+        emu.setMeta('TEBs', tebs)
+
+        seginfo = trace.getThreads()[trace.getMeta('ThreadId')]
+        if psize == 4:
+            emu.setSegmentInfo(e_i386.SEG_FS, seginfo, 0xffffffff)
+        elif psize == 8:
+            emu.setSegmentInfo(e_amd64.SEG_GS, seginfo, 0xffffffffffff)
+
+    return emu
+
+
+def vwFromTrace(trace, storagename='binary_workspace_from_vsnap.viv', filefmt=None, collapse=True, strict=True):
+    '''
+    Produce an envi emulator for this tracer object.
+
+    If filefmt is None, it will be auto-determined
+
+    If collapse, join adjacent maps
+    If strict, only join maps with the same permissions
+    '''
+    import vivisect
+    import envi.memory as e_memory
```

Comment:
We can probably move most of the imports in this function and the other to the top of the file.

Reply:
done